### PR TITLE
PR for waypoint_updater and twist_controller update

### PR DESCRIPTION
**Pull request by chriskcheung**
Hi Team, I updated waypoint_updater and twist_controller to handle deceleration base on traffic light and obstacle detection topics. Please review my changes and let me know if I missed anything or recommendation for improvement. Thanks.

### CODE REVIEW

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 1]
```diff
@@ -40,6 +42,8 @@ def __init__(self):
         rospy.Subscriber('/current_pose', PoseStamped, self.pose_callback)
         rospy.Subscriber('/base_waypoints', Lane, self.base_waypoints_callback)
         # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py]: remove?

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 2]
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
 
         :param index of the first waypoint.
         """
-        lane = Lane()
+        lane = self.get_final_lane(index)
         lane.header = self.base_waypoints.header
-        lane.waypoints = self.base_waypoints.waypoints[index:index + _NUM_WAYPOINTS_AHEAD]
+        #lane.waypoints = self.base_waypoints.waypoints[index:index + _NUM_WAYPOINTS_AHEAD]
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py]: remove?

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 3]
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
...
         self.final_waypoints_pub.publish(lane)
 
+    # update final Lane's waypoints base on traffic light or obstacle waypoint index
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py]: PEP8 style of function docstring.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 4]
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
...
+            # both traffic light and obstacle are detected
+            # find the closet waypoint among traffic light and obstacle waypoint index
+            elif self.traffic_light_wp_idx > self.obstacle_wp_idx:
+                target_wp_idx = self.traffic_light_wp_idx
+            else:
+                target_wp_idx = self.obstacle_wp_idx
+            lane.waypoints = self.declerate_waypoints(base_waypoints, closest_wp_idx, target_wp_idx)
+        return lane
+
+    # find decelerating waypoints
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py]: same here

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 5]
```diff
@@ -116,13 +178,13 @@ def base_waypoints_callback(self, base_waypoints):
         self.base_waypoints = base_waypoints
         rospy.loginfo('base_waypoints initialized')
 
-    def traffic_cb(self, msg):
+    def traffic_callback(self, data):
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py:190]: same here and below.

**Review by soniclue — APPROVED** [review]
just some code style stuff.

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (chriskcheung) [reply ros/src/waypoint_updater/waypoint_updater.py]: this is replaced with get_final_lane(index). I will clean up the commented line and resubmit.

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 3 above)
Comment (chriskcheung) [reply ros/src/waypoint_updater/waypoint_updater.py]: Will update it following PEP8 docstring

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 4 above)
Comment (chriskcheung) [reply ros/src/waypoint_updater/waypoint_updater.py]: Noted

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 5 above)
Comment (chriskcheung) [reply ros/src/waypoint_updater/waypoint_updater.py:190]: Noted

**Review by chriskcheung — COMMENTED** [review]
I move this out of base_waypoint for loop as it is static from input parameters

**Pull request merged**